runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,167 @@
+
+        // handle singular primitive and enum values immediately
+        if (strategy.descriptor.kind is PrimitiveKind) {
+            if (obj is Char) {
+                val value: Char = obj // without the explicit type annotation int value would be used
+                return value.toString()
+            }
+            if (obj is Long) {
+                return obj.toDouble()
+            }
+            return obj
+        }
+        if (strategy.descriptor.kind is UnionKind.ENUM_KIND) {
+            return obj.unsafeCast<Enum<*>>().name
+        }
+
+        val serializer = DynamicObjectEncoder(encodeDefaults)
+        serializer.encode(strategy, obj)
+        return serializer.result
+    }
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: T): dynamic = serialize(context.getContextualOrDefault(T::class), obj)
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: List<T?>): dynamic =
+        serialize(context.getContextualOrDefault(T::class).nullable.list, obj)
+}
+
+private class DynamicObjectEncoder(val encodeDefaults: Boolean) : ElementValueEncoder() {
+    private object NoOutputMark
+
+    var result: dynamic = NoOutputMark
+    lateinit var current: Node
+    var currentName: String = ""
+    var currentDescriptor: SerialDescriptor = StringDescriptor
+
+    class Node(val writeMode: WriteMode, val jsObject: dynamic) {
+        var index: Int = 0
+        lateinit var parent: Node
+    }
+
+    enum class WriteMode {
+        OBJ, MAP, LIST
+    }
+
+    override fun encodeElement(desc: SerialDescriptor, index: Int): Boolean {
+        current.index = index
+        currentDescriptor = desc
+
+        if (current.writeMode != WriteMode.MAP) {
+            currentName = desc.getElementName(index)
+        }
+        return true
+    }
+
+    override fun encodeValue(value: Any) {
+        val elementIsMapKey = current.writeMode == WriteMode.MAP && current.index % 2 == 0
+        if (elementIsMapKey) {
+            val elementDescriptor = currentDescriptor.getElementDescriptor(current.index)
+            currentName = if (elementDescriptor.kind is PrimitiveKind) {
+                value.toString()
+            } else if (elementDescriptor.kind is UnionKind.ENUM_KIND && value is Enum<*>) {
+                value.name
+            } else {
+                throw IllegalArgumentException(
+                    "Value of type ${elementDescriptor.name} can't be used in json as map key. " +
+                            "It should have either primitive or enum kind, but its kind is ${elementDescriptor.kind}.\n"
+                )
+            }
+        } else {
+            current.jsObject[currentName] = value
+        }
+    }
+
+    override fun encodeChar(value: Char) {
+        current.jsObject[currentName] = value.toString()
+    }
+
+    override fun encodeNull() {
+        if (encodeDefaults &&
```

Comment:
You don't need this check since `shouldEncodeElementDefault` is called by serializer _before_ any `encode` calls. So you should just override it: `override fun shouldEncodeElementDefault(desc: SerialDescriptor, index: Int): Boolean = encodeDefaults`

Reply:
Are you sure? Maybe I am doing something wrong but the shouldEncodeElementDefault method is not called at all if I omit the call here.
Do I need to rebase first?

Reply:
No, it's quite old functionality. Maybe you confused `class X(val s: String? = null)` vs `class Y(val s: String?)` – in case of Y, `shouldEncodeElementDefault` indeed would not be called. If you still want to drop nulls regardless whether they are optional or not, (which may confuse `DynamicObjectParser`, which tries to make the difference between property being `null` and missing (`undefined`)), then you indeed should apply custom logic here (which basically would be 'do not write null if encodeDefaults=true)

Reply:
Thanks for clearing things up. I now understand that the intended behaviour of `shouldEncodeElementDefault` is to omit properties that *exactly* have the default value. Which totally makes sense given its name.

IMO the intended behaviour of `shouldEncodeElementDefault` is almost never what I would want from `DynamicObjectSerializer`.

The original problem I wanted to solve was indeed to differentiate between undefined and null. I will add an optional property `omitNullProperties = false` and document it accordingly. We could still leave the optional `shouldEncodeElementDefault = true` and change the implementation to what you suggested above.